### PR TITLE
Fix #1323. lr_scheduler, clip_gradient, Xavier initializer are supported for R package

### LINKED ISSUE #1323: R-package lr_scheluder, clip_gradient and xavier initializer
I am trying to port a model from python to R, basically the model call in python is:
model = mx.model.FeedForward(
ctx = dev,
symbol = net,
num_epoch = 35,
learning_rate = 0.01,
momentum = 0.9,
wd = 0.0001,
clip_gradient = 5,
lr_scheduler = mx.lr_scheduler.FactorScheduler(step=epoch_size * lr_factor_epoch, factor = 0.1),
initializer = mx.init.Xavier(factor_type="in", magnitude=2.34))

but I can't find in R the equivalents for lr_scheduler (and mx.lr_scheduler.FactorScheduler) , clip_gradient and mx.init.Xavier. 
I guess they have not been implemented in R yet, right?

### PR DESCRIPTION
**Pull request by Puriney**
**How**

The Xavier initializer is borrowed from python implementation. This PR enables R package support `factor_type = c('in', 'out', 'avg')` and `rnd_type = c('gaussian', 'uniform')` to generate weights.

`clip_gradient` is implemented via `pmax`/`pmin` R built-in function. 

`lr_scheduler` is referring to python implementation `python/lr_scheduler.py`.  I noticed there is some discussion among your developer team and hopefully my PR makes it as well. If @thirdwing  could revise this PR, it will be appreciated ;-)

**What**

Give a try using the official example:

```r
# Using regerssion as example to test my xavier initializer

# Data
data(BostonHousing, package="mlbench")
train.ind = seq(1, 506, 3)
train.x = data.matrix(BostonHousing[train.ind, -14])
train.y = BostonHousing[train.ind, 14]
test.x = data.matrix(BostonHousing[-train.ind, -14])
test.y = BostonHousing[-train.ind, 14]

# Computing Symbols
data <- mx.symbol.Variable("data")
fc1 <- mx.symbol.FullyConnected(data, num_hidden=1)
lro <- mx.symbol.LinearRegressionOutput(fc1)

# Training model
print("Default training")
mx.set.seed(0)
model <-
  mx.model.FeedForward.create(lro, X=train.x, y=train.y,
                              ctx=mx.cpu(), num.round=50, array.batch.size=20,
                              learning.rate=2e-6, momentum=0.9,
                              eval.metric=mx.metric.rmse)


# Prediction using test data
preds <- predict(model, test.x)
# See diff
diff <- sqrt(mean((preds-test.y)^2))

# Training model with Xavier
print("Traing with Xavier initializer")
mx.set.seed(0)
modelX <-
  mx.model.FeedForward.create(lro, X=train.x, y=train.y,
                              ctx=mx.cpu(), num.round=50, array.batch.size=20,
                              learning.rate=2e-6, momentum=0.9,
                              eval.metric=mx.metric.rmse,
                              initializer = mx.init.Xavier(
                                rnd_type = "gaussian",
                                factor_type = "in",
                                magnitude = 2))
predsX <- predict(modelX, test.x)
diffX <- sqrt(mean((predsX - test.y)^2))

# Training model with clipping on gradient
print("Training model with clip")
mx.set.seed(0)
modelC <-
  mx.model.FeedForward.create(lro, X=train.x, y=train.y,
                              ctx=mx.cpu(), num.round=50, array.batch.size=20,
                              learning.rate=2e-6, momentum=0.9,
                              eval.metric=mx.metric.rmse,
                              clip_gradient = 1)
predsC <- predict(modelC, test.x)
diffC <- sqrt(mean((predsC - test.y)^2))

# Training model with clipping on gradient
print("Training model with schedule")
mx.set.seed(0)
modelS <-
  mx.model.FeedForward.create(lro, X=train.x, y=train.y,
                              ctx=mx.cpu(), num.round=50, array.batch.size=20,
                              learning.rate=2e-6, momentum=0.9,
                              eval.metric=mx.metric.rmse,
                              lr_scheduler=FactorScheduler(10, 0.9))

predsS <- predict(modelS, test.x)
diffS <- sqrt(mean((predsS - test.y)^2))

print(diff)
print(diffX)
print(diffC)
print(diffS)
```

```
[1] 7.800502
[1] 7.888685
[1] 16.15398
[1] 8.324275
```

Cc: @thirdwing  as I am interested in participating your GSOC 2016 project.

### CODE REVIEW

**Comment by thirdwing** [comment]
LGTM. And there might be some minor changes you can make.

(1) please add your name to https://github.com/dmlc/mxnet/blob/master/CONTRIBUTORS.md

(2) what about adding the example into documents?

@Gelu74 , can you help to test it?

PS: it is not **my** GSOC project. We can move the discussion to emails.

**Comment by Gelu74** [comment]
Thanks, yes, I'll give it a test and try to add an example for image classification

**Comment by terrytangyuan** [comment]
Good job on first PR! I'll take a closer look when I get a chance.

File: R-package/R/lr_scheduler.R [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+FactorScheduler <- function(step, factor_val) {
+  function(optimizerEnv){
+    num_update <- optimizerEnv$num_update
+    count      <- optimizerEnv$count
+    lr         <- optimizerEnv$lr
+
+    if (num_update > count + step){
+      count = count + step
+      lr = lr * factor_val
+      print(paste("Update[", num_update, "]: learning rate is changed to", lr, "\n"))
```
Comment (thirdwing) [inline R-package/R/lr_scheduler.R]: Change this line into `cat(paste("Update[", num_update, "]: learning rate is changed to ", lr, "\n", sep = ""))`

then I will merge it.
